pubtools/_pulp/tasks/publish.py

```diff
@@ -0,0 +1,221 @@
+            "--force", help="execute a full repo publish", type=bool, default=None
+        )
+        self.parser.add_argument(
+            "--published-before",
+            help="publish the repos last published before given date e.g. 2019-08-21",
+            type=publish_date,
+            default=None,
+        )
+        self.parser.add_argument(
+            "--url-regex",
+            help="publish repos whose repo url match",
+            default=None,
+            type=re.compile,
+        )
+
+    def run(self):
+        to_await = []
+        LOG.info("run publish")
+
+        # get repos applying filters
+        repos = self.get_repos()
+
+        # publish the repos found
+        publish_fs = self.publish(repos)
+
+        # wait for the publish to complete before
+        # flushing caches.
+        f_sequence(publish_fs).result()
+
+        # flush UD cache
+        to_await.extend(self.flush_ud(repos))
+
+        # flush CDN cache
+        to_await.extend(self.flush_cdn(repos))
+
+        # wait for everthing to finish.
+        for f in to_await:
+            f.result()
+
+        LOG.info("Publishing repositories completed")
+
+    @step("Publish")
+    def publish(self, repos):
+        out = []
+
+        publish_opts = PublishOptions(force=self.args.force, clean=self.args.clean)
+        for repo in repos:
+            LOG.info("Publishing %s", repo.id)
+            f = repo.publish(publish_opts)
+            f = f_map(f, partial(self.log_publish, repo))
+            out.append(f)
+
+        return out
+
+    def log_publish(self, repo, tasks):
+        # logs publish status
+        for task in tasks:
+            LOG.info(
+                "Publishing %s %s",
+                repo.id,
+                "successful" if task.succeeded else "unsuccessful",
+            )
+
+    @step("Flush UD cache")
+    def flush_ud(self, repos):
+        client = self.udcache_client
+        if not client:
+            LOG.info("UD cache flush is not enabled.")
+            return []
+
+        out = []
+        for repo in repos:
+            out.append(client.flush_repo(repo.id))
+            if repo.eng_product_id:
+                out.append(client.flush_product(repo.eng_product_id))
+
+        return out
+
+    @step("Flush CDN cache")
+    def flush_cdn(self, repos):
+        if not self.fastpurge_client:
+            LOG.info("CDN cache flush is not enabled.")
+            return []
+
+        def purge_repo(repo):
+            to_flush = []
+            for url in repo.mutable_urls:
+                flush_url = os.path.join(
+                    self.fastpurge_root_url, repo.relative_url, url
+                )
+                to_flush.append(flush_url)
+
+            LOG.debug("Flush: %s", to_flush)
+            flush = self.fastpurge_client.purge_by_url(to_flush)
+            return f_map(flush, lambda _: repo)
+
+        return [purge_repo(r) for r in repos if r.relative_url]
+
+    @step("Get repos")
```

Comment:
It seems from the user's point of view, this step has the same meaning as "Check repos" in clear-repo command, so could it please be named the same?

If you think this step name is better, feel free to change the one in clear-repo.

Reply:
I think "get repos" made more sense as this step actually gets the repos either for the repo ids provided or the repo ids fetched applying the filters. Same applies to clear-repo too, so updated it there.

Reply:
hm, but except get repos, you also checked if any repos' missing?

Reply:
I think that "check repos" is probably going to make more sense to the end-users.  Think about if you put a bank card in an ATM.  While that's loading data, is it more likely to say "Getting your account..." or "Checking your account..." ?  

It should be the latter, because the fact that some representation of the account/repo is being "gotten" from one place to another is an implementation detail which only makes sense if you think of the individual components of the system (i.e. that an ATM / Pub worker exists as one component which must "get" info from some separate component).  Ideally, users shouldn't be confronted with such details and it should appear more like one coherent system.

But, I mainly care about consistency, so I don't mind as long as the step is named the same in different tasks.